xls/modules/dbe/docs/README.md

```diff
@@ -0,0 +1,289 @@
+      that symbol yet, and it must be emitted at some point, and also
+      this symbol may be able to start a new matching string on its own).
+  * Otherwise:
+    * __(17)__ Emit an *UNMATCHED_SYMBOL* token for the current symbol.
+    * Go to step __(1)__.
+
+## DSLX implementation
+
+LZ4 encoder in DSLX is implemented as an FSM-like which changes states at most
+once per "tick". It has a form of a _proc_ module.
+
+Compared to the reference flowchart depicted above, this FSM is a bit more
+complicated as it has to take into account some corner cases:
+* Prefilling the FIFO with data before beginning the processing.
+* Correctl handling of current symbol when it runs into an EOF.
+* Draining the FIFO after EOF has been observed.
+* Allowing "warm" restart of the algorithm after EOF to allow compression of
+  *dependent* blocks - that is, without resetting the HT and the HB between the
+  blocks (this allows tokens from the curent block to refer HB symbols of the
+  previous block).
+* System-level integration considerations to allow this block to be chained
+  with other XLS-based data pre-/postprocessors:
+  * *MATCHED_SYMBOL* tokens. Whenever the encoder adds a symbol to the matching
+    string, it emits a *MATCHED_SYMBOL* token, which can be used by the
+    postprocessing blocks to reconstruct symbols encoded by *MATCH* tokens
+    without having to decode *MATCH*es (such decoding would require a full-size
+    history buffer RAM). These tokens should not be stored into the final
+    encoded block.
+  * Support for a limited form of in-band control signaling via
+    *marker tokens*:
+    * *END* marker that signals end of block (EOF condition).
+    * *ERROR_* family of markers that allow passing error codes between
+      processing blocks. These tokens abort the encoding.
+    * *RESET* marker that performs a *cold* restart of the encoder (clearing HT
+      RAM) and other blocks in the chain, also clearing all error conditions.
+
+One example of a post-processor module can be a *block writer* proc that
+gathers tokens produced by the encoder and encodes them using a standardized
+byte-oriented *LZ4 Block Format*. Implementation of alternative encoding
+schemes may be of interest as well, as the header format used by the standard
+LZ4 requires one to know the number of unmatched symbols between two matches
+before those symbols are emitted, making bufferless stream processing
+difficult.
+
+### Data format
+
+#### PlainData
+
+The encoder consumes a stream of raw symbols, intermixed with control markers.
+This is represented using a parametrized `PlainData` DSLX structure:
+```rust
+pub struct PlainData<DATA_WIDTH: u32> {
+    is_marker: bool,
+    data: uN[DATA_WIDTH],
+    mark: Mark,
+}
+```
+
+* *is_marker* tells whether this object is a symbol or a marker.
+* *data* communicates a symbol whenever *is_marker* is not set.
+* *mark* communicates a control mark whenever *is_marker* is set.
+
+#### Token
```

Comment:
I'd suggest writing another section on how these tokens can be converted into a stream compatible with the LZ4 block format (https://github.com/lz4/lz4/blob/dev/doc/lz4_Block_format.md)

Notably the following constraints
  1. That there is a minimum match length of 4 bytes
  2. That after each literal section there MUST be a minimum match
  3. The end of the sequence must have 5 literal bytes
  4. The last match must have an offset >= 12

Reply:
I will do this. In short, the idea is that you can use data conveyed within "MATCHED_SYMBOLs" to rewrite "MATCHes", e.g. if your encoding format requires a match of at least 4 symbols, and the encoder emits a match of 2 symbols, there will be two MATCHED_SYMBOL tokens accompanying a MATCH, so the postprocessor can remove the match and convert MATCHED_SYMBOLs into UNMATCHES_SYMBOLs (i.e. literals).

This way it is possible to remove or shorten matches, and I think it should be enough to cover all the requirements imposed by LZ4 standard.

Reply:
Thanks -- I think the discussion on token format is relevant here as well.  Is it even possible to do such a thing without needing to buffer the entire block?

Reply:
I've added a description to the doc that shows how matches can be rewritten to fulfill these requirements.

One note, though, about requirement 2: *That after each literal section there MUST be a minimum match* - a match must be there, except if it's the end of the block. The end of the block must contain only literals and no match. Thus, a typical LZ4 block looks like this:
```
literals - match - literals - match - literals - END
```

One other requirement is about the minimal match size, and that's easy to implement with only 4 (or even 3) bytes of FIFO memory. The rest of requirements discuss how exactly that last sequence of "literals" has to look like. Those requirements can be fulfilled by making sure that there is no match within the last 12 symbols in the block - which is as simple as "buffering up to 12 symbols and if there are any matches in them, replacing those matches with literals". The tricky part is if the block ends in a long match that produces more than those 12 symbols. In this case, the match has to be split - and it's possible to do this without needing more than a 12-symbol buffer (or, how the doc says, 15-token buffer, since in reality we also want to buffer match tokens, there can be up to 3 of those per 12 symbols).